File: docusign/webhooks.js
// docusign/webhooks.js
// ─────────────────────────────────────────────────────────────────────────────
// DocuSign Connect (webhooks): DocuSign POSTs to your endpoint whenever an
// envelope's status changes, so you don't have to poll. This module verifies the
// payload's HMAC signature and normalizes it into a small, stable event shape.
//
// Configure in DocuSign Admin → Connect: point a custom configuration at
//   <PUBLIC_URL>/api/docusign/webhook
// with "Include HMAC Signature" enabled and the same secret as DOCUSIGN_WEBHOOK_SECRET.
// ─────────────────────────────────────────────────────────────────────────────

import crypto from 'node:crypto';
import { config } from './config.js';

/**
 * Verify a Connect HMAC signature (header: X-DocuSign-Signature-1).
 * Returns true when no secret is configured (verification disabled) so dev/sim
 * webhooks still flow; set DOCUSIGN_WEBHOOK_SECRET in production to enforce it.
 * @param {Buffer|string} rawBody  the EXACT bytes DocuSign sent.
 * @param {string} signatureHeader
 */
export function verifySignature(rawBody, signatureHeader) {
  if (!config.webhookSecret) return true;
  if (!signatureHeader) return false;
  const computed = crypto.createHmac('sha256', config.webhookSecret).update(rawBody).digest('base64');
  const a = Buffer.from(computed);
  const b = Buffer.from(String(signatureHeader));
  return a.length === b.length && crypto.timingSafeEqual(a, b);
}

/**
 * Normalize a Connect payload into { envelopeId, status, recipients, raw }.
 * Supports both the modern JSON ("aggregate") shape and the simpler flat shape.
 */
export function parseEvent(body) {
  if (!body || typeof body !== 'object') return null;

  // Modern DocuSign Connect JSON: { event, data: { envelopeId, envelopeSummary: {...} } }
  if (body.data?.envelopeId || body.data?.envelopeSummary) {
    const summary = body.data.envelopeSummary || {};
    return {
      envelopeId: body.data.envelopeId || summary.envelopeId,
      status: (summary.status || body.event || '').toLowerCase().replace(/^envelope-/, ''),
      completedAt: summary.completedDateTime || null,
      recipients: (summary.recipients?.signers || []).map((s) => ({ email: s.email, name: s.name, status: s.status, signedAt: s.signedDateTime || null })),
      raw: body,
    };
  }

  // Flat shape: { envelopeId, status, ... }
  if (body.envelopeId) {
    return {
      envelopeId: body.envelopeId,
      status: String(body.status || '').toLowerCase(),
      completedAt: body.completedDateTime || body.completedAt || null,
      recipients: body.recipients || [],
      raw: body,
    };
  }
  return null;
}

File: backend/app/docusign/webhooks.py
"""DocuSign Connect (webhooks): verify HMAC + normalize the event shape."""
import base64
import hashlib
import hmac

from .config import config


def verify_signature(raw_body: bytes | str, signature_header: str | None) -> bool:
    # Fail closed: with no configured secret we cannot authenticate the caller,
    # so reject rather than trust an unverifiable webhook.
    if not config["webhookSecret"]:
        return False
    if not signature_header:
        return False
    body = raw_body.encode() if isinstance(raw_body, str) else (raw_body or b"")
    computed = base64.b64encode(hmac.new(config["webhookSecret"].encode(), body, hashlib.sha256).digest()).decode()
    return hmac.compare_digest(computed, str(signature_header))


def parse_event(body: dict | None) -> dict | None:
    if not body or not isinstance(body, dict):
        return None

    data = body.get("data")
    if isinstance(data, dict) and (data.get("envelopeId") or data.get("envelopeSummary")):
        summary = data.get("envelopeSummary") or {}
        status = str(summary.get("status") or body.get("event") or "").lower()
        if status.startswith("envelope-"):
            status = status[len("envelope-"):]
        signers = ((summary.get("recipients") or {}).get("signers")) or []
        return {
            "envelopeId": data.get("envelopeId") or summary.get("envelopeId"),
            "status": status,
            "completedAt": summary.get("completedDateTime"),
            "recipients": [{"email": s.get("email"), "name": s.get("name"), "status": s.get("status"), "signedAt": s.get("signedDateTime")} for s in signers],
            "raw": body,
        }

    if body.get("envelopeId"):
        return {
            "envelopeId": body.get("envelopeId"),
            "status": str(body.get("status") or "").lower(),
            "completedAt": body.get("completedDateTime") or body.get("completedAt"),
            "recipients": body.get("recipients") or [],
            "raw": body,
        }
    return None
